### PR TITLE
[JCLOUDS-873] fixed

### PR DESCRIPTION
**Pull request by fmartelli**

### CODE REVIEW

**Comment by fmartelli** [comment]
@nacx, @andreaturli could you review this PR?

The PR has been rebased on the current latest commit (aac957447270e6bc78d62de628cfb539add5045f).

Mock and live tests for network security group operations have been provided.
Several changes have been required both to improve/fix live tests and to include recently merged PRs.
 
All the changes have been squashed in a single commit.

Thank you for your support.
Regards.

**Comment by nacx** [comment]
Hi @fmartelli!
Now that the other Azure PRs have been merged, could you rebase this one so it only contains the commits that are relevant to the JIRA issue? I look forward to having a look at this and merging it!
Thanks!

**Comment by fmartelli** [comment]
Hi @nacx, I'm working o this issue. Live tests take more then 1h and 30m .... stay tuned.
Thanks.

**Comment by fmartelli** [comment]
Hi @nacx, I've just pushed the rebase  and the squash into just one commit.

Please, consider that the commit includes all required changes for this issue: 
1. completes Network Security Group API implementation and adds mock and live tests
2. fix test conflicts inherited from several rebases
3. add a required improvement of ConflictManagementPredicate

Thank you for your support.

**Comment by fmartelli** [comment]
@nacx, I'm missing why you are referencing #161 from #163.
#161 is completed and ready to be reviewed/merged. Please let me know if you need something more.

Thanks.

**Comment by nacx** [comment]
Oh, apologies. I didn't notice both refer to a different issue and though 163 was this one rebased. I didn't look at the content of the commits. Will review it now :)

File: azurecompute/src/test/java/org/jclouds/azurecompute/features/NetworkSecurityGroupApiLiveTest.java [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+      assertTrue(new ConflictManagementPredicate(api) {
+
+         @Override
+         protected String operation() {
+            return api().removeFromSubnet(VIRTUAL_NETWORK_NAME, DEFAULT_SUBNET_NAME, GROUP1);
+         }
+      }.apply(GROUP1));
+   }
+
+   @AfterClass
```
Comment (nacx) [inline azurecompute/src/test/java/org/jclouds/azurecompute/features/NetworkSecurityGroupApiLiveTest.java]: Use `@AfterClass(alwaysRun = true)`

File: azurecompute/src/test/java/org/jclouds/azurecompute/features/NetworkSecurityGroupApiMockTest.java [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+         assertSent(server, "PUT", "/services/networking/networksecuritygroups/mygroup/rules/myrule");
+      } finally {
+         server.shutdown();
+      }
+   }
+
+   private NetworkSecurityGroupApi networkSecurityGroupApi(final MockWebServer server) {
+      return api(server.getUrl("/")).getNetworkSecurityGroupApi();
+   }
+}
```
Comment (nacx) [inline azurecompute/src/test/java/org/jclouds/azurecompute/features/NetworkSecurityGroupApiMockTest.java]: Add the mock tests that exercise the defined fallbacks.

**Comment by nacx** [comment]
Just a couple of comments, but it looks good! Thanks @fmartelli!

Just to be sure: the ConflictManagement predicate is basically used to deal with the Azure operations that block the resources error and wait until the API allows the new incoming operations, right?

**Comment by fmartelli** [comment]
exactly: it has been required to manage conflict errors (http status code 409)

**Comment by fmartelli** [comment]
Rebased on 11ae78af457c298f4d4771c8a4326551c13c401c.

**Comment by fmartelli** [comment]
@nacx, just rebased again on the latest commit.
Any news?

**Comment by nacx** [comment]
Thanks! Just waiting for the tests mentioned in https://github.com/jclouds/jclouds-labs/pull/161#discussion_r28064957.

**Comment by fmartelli** [comment]
@nacx, I'm sorry for the mistake.
I've just pushed the fix requested by you.
Thanks.

**Comment by nacx** [comment]
Squashed and pushed to [master](http://git-wip-us.apache.org/repos/asf/jclouds-labs/commit/29435e1c) and [1.9.x](http://git-wip-us.apache.org/repos/asf/jclouds-labs/commit/2b69bed1). Thanks @fmartelli!

**Pull request closed without merging**